rust/wasmtime_bindings/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+}
+
+fn create_module(engine: &mut Engine, script: &str) -> Result<Box<Module>> {
+    match wasmtime::Module::new(&engine.wasmtime_engine, script) {
+        Ok(module) => Ok(Box::new(Module {
+            wasmtime_module: module,
+        })),
+        Err(e) => Err(anyhow!("Compilation failed: {}", e)),
+    }
+}
+
+pub struct Store {
+    wasmtime_store: wasmtime::Store<wasmtime_wasi::WasiCtx>,
+}
+
+fn create_store(engine: &mut Engine) -> Result<Box<Store>> {
+    let wasi = wasmtime_wasi::tokio::WasiCtxBuilder::new().build();
+    let mut store = wasmtime::Store::new(&mut engine.wasmtime_engine, wasi);
+    // TODO: make this configurable
+    store.out_of_fuel_async_yield(10000, 4096);
+    Ok(Box::new(Store {
+        wasmtime_store: store,
+    }))
+}
+
+pub struct Memory {
+    wasmtime_memory: wasmtime::Memory,
+}
+
+impl Memory {
+    fn data(self: &Memory, store: &Store) -> *mut u8 {
+        self.wasmtime_memory.data_ptr(&store.wasmtime_store)
+    }
+    fn size(self: &Memory, store: &Store) -> u64 {
+        self.wasmtime_memory.size(&store.wasmtime_store)
+    }
+    fn grow(self: &mut Memory, store: &mut Store, delta: u64) -> Result<u64> {
+        self.wasmtime_memory.grow(&mut store.wasmtime_store, delta)
+    }
+}
+
+fn get_memory(instance: &Instance, store: &mut Store) -> Result<Box<Memory>> {
+    match instance
+        .wasmtime_instance
+        .get_export(&mut store.wasmtime_store, "memory")
+    {
+        None => Err(anyhow!(
+            "Memory export not found - please export `memory` in the wasm module"
+        )),
+        Some(export) => match export.into_memory() {
+            None => Err(anyhow!(
+                "Exported object memory is not a WebAssembly memory"
+            )),
+            Some(memory) => Ok(Box::new(Memory {
+                wasmtime_memory: memory,
+            })),
+        },
+    }
+}
+
+fn get_abi(instance: &Instance, store: &mut Store, memory: &Memory) -> Result<u32> {
+    match instance
+        .wasmtime_instance
+        .get_export(&mut store.wasmtime_store, "_scylla_abi")
+    {
+        None => Err(anyhow!(
+            "ABI version export not found - please export `_scylla_abi` in the wasm module"
+        )),
+        Some(export) => match export.into_global() {
+            None => Err(anyhow!("Exported object _scylla_abi is not a global")),
+            Some(global) => match global.get(&mut store.wasmtime_store) {
+                wasmtime::Val::I32(x) => {
+                    let mut bytes = [0u8; 4];
+                    memory
+                        .wasmtime_memory
+                        .read(&store.wasmtime_store, x as usize, &mut bytes)?;
+                    Ok(u32::from_le_bytes(bytes))
+                }
+                _ => Err(anyhow!("Exported global _scylla_abi is not an i32")),
+            },
+        },
+    }
+}
+
+pub struct Engine {
+    wasmtime_engine: wasmtime::Engine,
+}
+
+fn create_engine() -> Result<Box<Engine>> {
```

Comment:
am I right to assume that Result<T> gets translated by cxx to produce a function that returns a T if the result is Ok, and throws an exception in case of Err?

Reply:
Yup: https://cxx.rs/binding/result.html#returning-result-from-rust-to-c